win32ss/gdi/eng/mdevobj.c

```diff
@@ -0,0 +1,202 @@
+            if (!NT_SUCCESS(Status))
+            {
+                /* Ignore this device and continue */
+                ERR("Failed to open VIDEO key: status 0x%08x\n", Status);
+                continue;
+            }
+            cbSize = sizeof(DeviceKey);
+            Status = RegQueryValue(hKey,
+                          pGraphicsDevice->szNtDeviceName,
+                          REG_SZ,
+                          DeviceKey,
+                          &cbSize);
+            ZwClose(hKey);
+            if (!NT_SUCCESS(Status))
+            {
+                /* Ignore this device and continue */
+                ERR("Failed to open get device key for '%S': status 0x%08x\n", pGraphicsDevice->szNtDeviceName, Status);
+                continue;
+            }
+            Status = RegOpenKey(DeviceKey, &hKey);
+            if (!NT_SUCCESS(Status))
+            {
+                /* Ignore this device and continue */
+                ERR("Failed to open open device key '%S' for '%S': status 0x%08x\n", DeviceKey, pGraphicsDevice->szNtDeviceName, Status);
+                continue;
+            }
+#define READ(field, str, flag) \
+    if (RegReadDWORD(hKey, L##str, &dwValue)) \
+    { \
+        dmDefault.field = dwValue; \
+        dmDefault.dmFields |= flag; \
+    }
+            READ(dmBitsPerPel, "DefaultSettings.BitsPerPel", DM_BITSPERPEL);
+            READ(dmPelsWidth, "DefaultSettings.XResolution", DM_PELSWIDTH);
+            READ(dmPelsHeight, "DefaultSettings.YResolution", DM_PELSHEIGHT);
+            READ(dmDisplayFlags, "DefaultSettings.Flags", DM_DISPLAYFLAGS);
+            READ(dmDisplayFrequency, "DefaultSettings.VRefresh", DM_DISPLAYFREQUENCY);
+            READ(dmPanningWidth, "DefaultSettings.XPanning", DM_PANNINGWIDTH);
+            READ(dmPanningHeight, "DefaultSettings.YPanning", DM_PANNINGHEIGHT);
+            READ(dmDisplayOrientation, "DefaultSettings.Orientation", DM_DISPLAYORIENTATION);
+            READ(dmDisplayFixedOutput, "DefaultSettings.FixedOutput", DM_DISPLAYFIXEDOUTPUT);
+            READ(dmPosition.x, "Attach.RelativeX", DM_POSITION);
+            READ(dmPosition.y, "Attach.RelativeY", DM_POSITION);
+            RegReadDWORD(hKey, L"Acceleration.Level", &dwAccelerationLevel);
+            ZwClose(hKey);
+        }
+
+        /* Get or create a PDEV for these settings */
+        if (LDEVOBJ_bProbeAndCaptureDevmode(pGraphicsDevice, pdm ? pdm : &dmDefault, &localPdm, !pdm))
+        {
+            ppdev = PDEVOBJ_Create(pGraphicsDevice, localPdm, dwAccelerationLevel, LDEV_DEVICE_DISPLAY);
+        }
+        else
+            ppdev = NULL;
+
+        if (ppdev)
+        {
+            /* Great. We have a found a matching PDEV. Store it in MDEV */
+            TRACE("Adding '%S' to MDEV %p\n", pGraphicsDevice->szWinDeviceName, pmdev);
+            PDEVOBJ_vReference(ppdev);
+            pmdev->dev[pmdev->cDev].ppdev = ppdev;
+            pmdev->cDev++;
```

Comment:
Question: are all these done under a lock?

Reply:
Not needed, because pmdev is some memory allocated in this function, and filled here before being returned